starlite/datastructures/provide.py

```diff
@@ -95,21 +88,30 @@ def add(self, generator: Union[Generator[Any, None, None], AsyncGenerator[Any, N
         self._generators.append(generator)
 
     @staticmethod
-    def _wrap_next(generator: "AnyGenerator") -> Callable[[], Any]:
+    def _wrap_next(generator: "AnyGenerator") -> Callable[[], Coroutine[None, None, None]]:
         if isasyncgen(generator):
 
             async def wrapped_async() -> None:
-                await anext(generator, None)  # type: ignore[arg-type]
+                await async_next(generator, None)  # type: ignore[arg-type]
 
             return wrapped_async
 
         def wrapped() -> None:
             next(generator, None)  # type: ignore[arg-type]
 
-        return wrapped
+        return AsyncCallable(wrapped)
 
-    def to_background_tasks(self) -> List[BackgroundTask]:
-        return [BackgroundTask(self._wrap_next(gen)) for gen in self._generators]
+    def wrap_asgi(self, app: "ASGIApp") -> "ASGIApp":
+        async def wrapped(scope: "Scope", receive: "Receive", send: "Send") -> None:
+            if len(self._generators) == 1:
+                await self._wrap_next(self._generators[0])()
+            elif self._generators:
+                async with anyio.create_task_group() as tg:
```

Comment:
```suggestion
                async with create_task_group() as task_group:
```

I know im annoying, sorry.

Reply:
No worries, I like me a thorough review :heart: